### ISSUE TITLE
Generator does not typecheck if the inferred return type is not a subtype of the declared return type

### ISSUE THREAD

**Issue by yacinehmito** [microsoft/TypeScript#33633]
**TypeScript Version:**  3.7.0-dev.20190927
(also tested with 3.6.3)

**Search Terms:**

Generator, generators, inference

**Code**

```ts
type ExampleReturnType = {
  keyWithLiteralValue: 'foo' | 'bar';
};

function normalFunctionThatTypechecks(): ExampleReturnType {
  return {
    // TS knows that `keyWithLiteralValue` has type `'foo' | 'bar'`
    keyWithLiteralValue: 'foo',
  };
}

function* generatorFunctionThatDoesNotTypecheck(): Generator<
  undefined,
  ExampleReturnType
> {
  yield undefined;
  return {
    // TS thinks that `keyWithLiteralValue` has type `string`
    // so it refuses to typecheck
    keyWithLiteralValue: 'foo',
  }
}
```

A repo with this example is available at https://github.com/yacinehmito/typescript-issue-generator-inference

**Expected behavior:**

I expected the example to typecheck.

**Actual behavior:**

The example cannot typecheck because TypeScript infers the return type as `{ keyWithLiteralValue: string }`, and then tries to unify with `{ keyWithLiteralValue: 'foo' | 'bar' }` which has to fail.

**Playground Link:** [Here](https://www.typescriptlang.org/play/?ts=Nightly#code/C4TwDgpgBAogHgQwLZgDYQEoWAVwE4B2AKuNALxQDeAUFFANYQgDqAlsABYAy7EeCqAGoCcEAFxQA5ADMA9rMlQAPlIBGCPJIDc1AL47q0nAQDGwVrIJQCsvEgEAxY2YvEOCYCUgmOEE-QBnAAoASgl4ZDRMbHxiUipaKDwYwgS6OgB6DKgiAGUGGwB3AKhODygAA0YWdm5efiERCAqodxLQSEqZeUUVSXVNCsS6arZOHmA+AWFUUQluhQAaRP09akNnc0sAKigAcwgCKeBbJ1Mttw8AEVkIAIA5WU9SHz96UIkAcUPj2wAeRLGAAmEGkrCOQOWdAiKHQWFwhC8EGoAD40lAQKwIKggVBgaDwRAgTo6MkEVYaOkoFkcvlOODAqV3MBKqNahMpo1Zs1Wgh2vEKgFgHhwXshlSaQFZFB2ElQTgAndStKOn5fP5hgwmGM6pMGjM5lI5EsVmsgA)

**Related Issues:** May be related to this: https://github.com/microsoft/TypeScript/issues/32728

**Comment by yacinehmito** [comment]
Polite "bump" (I don't even know if it does anything).

I don't really mind the bug personally because I'm not using generators yet but I think it is a concerning issue. Would somebody be kind enough to look it up?

**Comment by samyue** [comment]
Not checking the return type causes issues when I try to refactor code by renaming one property of the return type with IDE. The typescript doesn't pickup the issue if the property has wrong name. 

Related issue: https://github.com/microsoft/TypeScript/issues/241

**Comment by DanielRosenwasser** [comment]
It looks like this is potentially fixed?

**Comment by yacinehmito** [comment]
Hello. I can confirm it has been fixed indeed in TypeScrip 4.0.2, per the playground link.
Thanks you very much!
